ipn/ipnlocal/local.go

```diff
@@ -6449,13 +6449,36 @@ func (n lastSuggestedExitNode) asAPIType() (res apitype.ExitNodeSuggestionRespon
 	return res, ErrUnableToSuggestLastExitNode
 }
 
+// processAllowedCandidates converts a string slice of exit node candidate IDs to a map of strings and ints. Int value represents nothing and won't be used.
+// Used for a faster lookup comparison in suggestExitNode.
+func processAllowedCandidates(allowedCandidates []string) (candidates map[string]int) {
+	candidates = make(map[string]int, len(allowedCandidates))
+	for _, candidate := range allowedCandidates {
+		candidates[candidate] = 0
+	}
+	return candidates
+}
+
 func suggestExitNode(report *netcheck.Report, netMap *netmap.NetworkMap, r *rand.Rand) (res apitype.ExitNodeSuggestionResponse, err error) {
 	if report.PreferredDERP == 0 {
 		return res, ErrNoPreferredDERP
 	}
+	var allowedCandidates map[string]int
+	if allowed, _ := syspolicy.GetStringArray(syspolicy.AllowedSuggestedExitNodes, []string{}); len(allowed) != 0 {
+		allowedCandidates = processAllowedCandidates(allowed)
+	}
 	candidates := make([]tailcfg.NodeView, 0, len(netMap.Peers))
 	for _, peer := range netMap.Peers {
-		if peer.CapMap().Has(tailcfg.NodeAttrSuggestExitNode) && tsaddr.ContainsExitRoutes(peer.AllowedIPs()) {
+		var isCandidate bool
+		if len(allowedCandidates) != 0 && peer.CapMap().Has(tailcfg.NodeAttrAutoExitNode) {
```

Comment:
For feature checking via capability, you're going to want `netMap.SelfNode.CapMap()` because that's set for the current node, not for peers.

I checked with @kabirsikand and it sounds like enforcement should just be the same as general MDM policy usage, so I don't think you need to check for that here. That will dramatically simplify this bit of code.

Reply:
good point agreed